### ISSUE TITLE
Incorporate protocol identifier into proof transcripts

### ISSUE THREAD

**Issue by marsella** [boltlabs-inc/tss-ecdsa#141]
Each protocol instantiation has a unique identifier. This should be incorporated into the transcripts used to generate challenges in zero-knowledge proofs, so that proofs cannot be reused across multiple protocol instantiations.

## Proposed solution
Once transcript initialization has been pulled outside of each proof, make the transcript a field on each subprotocol participant (alternately, it could be passed around to every protocol method, but that seems potentially messy to me). When the protocol participant is instantiated with a protocol identifier, put that identifier into the transcript. Use the transcript in proofs.

I think the most complicated issue with this will be that transcripts are relevant on a per-party basis -- e.g. if I'm running a protocol with parties A and B, the transcript I send to B should not have details of any previous proofs I sent to party A. Perhaps this means we instantiate the transcript with fixed variables known to all parties, then clone it for each proof (this would mean if I did two proofs with party A, the transcripts for the second would not have details of the first). Or, we could hold one transcript for each party we communicate with, and keep a running record of all proofs with that party.

## Related issues
Tentatively, this is blocked by:
- Defining protocol id more consistently for each participant #140 
- Pull out proof transcript initialization from proofs themselves? #49 

Related issue:
- #4 is a superset of this issue

**Comment by indomitableSwan** [comment]
This naming is very confusing. Is "protocol identifier" used here to mean "subsession identifier"? I would think a "protocol identifier" is an object that identifies a protocol, not a specific run of the said protocol.

**Comment by marsella** [comment]
Yeah, I switched to "subsession identifier" and intend to make a terminology list somewhere (maybe in #91).

**Comment by indomitableSwan** [comment]
#141 still talks about protocol identifiers, though. Is #141 the issue that should specify renaming to "subsession identifier"?

**Comment by marsella** [comment]
> #141 still talks about protocol identifiers, though. Is #141 the issue that should specify renaming to "subsession identifier"?

I rewrote #140 and #141 to specify subsession identifiers.

**Comment by Ra1issa** [comment]
Unfortunately we currently can't pass the session identifier `sid` to the `SharedContext` because `InnerProtocolParticipant` does not store `sid`. There is a similar issue with #241 and it seems that #290 needs to be addressed first before proceeding any further.
